pkg/kubelet/kubelet_pods.go

```diff
@@ -643,9 +643,17 @@ func (kl *Kubelet) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container
 				if len(envFrom.Prefix) > 0 {
 					k = envFrom.Prefix + k
 				}
-				if errMsgs := utilvalidation.IsEnvVarName(k); len(errMsgs) != 0 {
-					invalidKeys = append(invalidKeys, k)
-					continue
+
+				if utilfeature.DefaultFeatureGate.Enabled(features.RelaxedEnvironmentVariableValidation) {
```

Comment:
This seems to bring a risk, if the user closes the feature gate and has used relaxed validation before update (it can still pass top-level validation), but the start container will fail here.

Reply:
I think the feature gate really just controls admission.  There are a few archetype changes that need gates:

For changes which add a net-new API field, that means the REST code (pkg/registry/...) does:
```
if gate_disabled && !already_in_use {
    wipe the field
}
```
Validation and impl code don't really need to check the gate.

For changes which don't add a new field, but change what's allowed in the API (e.g. new enums, loosening validation), the validation code does:
```
if gate_disabled && !already_in_use {
    validation error
}
```
Impl code doesn't really need to check the gate.

For changes which don't change the API, but change how things operate, the impl code does:
```
if gate_enabled {
    do something new
} else {
    do something old
}
```

@liggitt I have been meaning to write this down ....somewhere.  Do you agree with the general positioning?  Currently it's done however people think makes sense, with no consistency.

Reply:
For this specific case, I am unsure.  Kubelet should not really need to re-validate pods it gets from the apiserver, but I guess it is possible for the apiserver to feed a value that kubelet does not understand.  More important is static pods - someone has to validate those.  Unfortunately the validation code is all internal-version, so not exactly reusable unless we want kubelet to use internal-versions.

I am not as familiar with this path.

@tallclair or @SergeyKanzhelev - what's the norm here?  How is it supposed to work?    

If we do not validation here, what is the failure mode - static pod sets `name: foo=bar` and runtime fails to create the pod?

Reply:
> For changes which add a net-new API field, that means the REST code (pkg/registry/...) does:
> 
> ```
> if gate_disabled && !already_in_use {
>     wipe the field
> }
> ```
> 
> Validation and impl code don't really need to check the gate.

I agree with that. That's roughly what is described in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version but it's *very* well buried.

> For changes which don't add a new field, but change what's allowed in the API (e.g. new enums, loosening validation), the validation code does:
> 
> ```
> if gate_disabled && !already_in_use {
>     validation error
> }
> ```

Rather than validation code checking the diff and feature gate directly, I would model this as validation options passed into the Validate() and ValidateUpdate() functions. That makes it easier to unit test and easier to centralize the decision about when to allow the new value in the strategy. That is also documented in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-enum-value-in-existing-field but could certainly improve its visibility and clarity.



> For changes which don't change the API, but change how things operate, the impl code does:
> 
> ```
> if gate_enabled {
>     do something new
> } else {
>     do something old
> }
> ```

Agreed.

Reply:
> More important is static pods - someone has to validate those. Unfortunately the validation code is all internal-version, so not exactly reusable unless we want kubelet to use internal-versions.

That's exactly what kubelet does:

https://github.com/kubernetes/kubernetes/blob/8674282a054d3ae32e2e009dab6f8a0da3689828/pkg/kubelet/config/common.go#L109-L191

Reply:
> If we do not validation here, what is the failure mode

CRI accepts a wider range of characters as env names than k8s, and if we just don't validate the envvar name here, it shouldn't fail here

Since we have validated the static pod using the [validation code](https://github.com/kubernetes/kubernetes/blob/19df52c41a25cedb977b1dc064cdaa916261de67/pkg/kubelet/config/common.go#L110-L142), I think we can use the `IsRelaxedEnvVarName` function here to validate it?

I will seek the help of the SIG Node maintainers at slack, their advice would definitely be better...

Reply:
So static pods is one case and I've seen mentions in the code of standalone Kubelet. I don't really know how much that is used (maybe just for static pods). So I think validation in kubelet makes sense.

I think reusing the function makes sense to me.

Reply:
> I think are suggesting setting the feature gate as an option for validation and removing the feature gates from validation logic.

I don't quite understand here, may you elaborate?

Reply:
You should not have feature gates in any validation code and just instead set the PodValidationOptions.

See:

```
// PodValidationOptions contains the different settings for pod validation
type PodValidationOptions struct {
	// Allow invalid pod-deletion-cost annotation value for backward compatibility.
	AllowInvalidPodDeletionCost bool
	// Allow invalid label-value in LabelSelector
	AllowInvalidLabelValueInSelector bool
	// Allow pod spec to use non-integer multiple of huge page unit size
	AllowIndivisibleHugePagesValues bool
	// Allow pod spec to use status.hostIPs in downward API if feature is enabled
	AllowHostIPsField bool
	// Allow invalid topologySpreadConstraint labelSelector for backward compatibility
	AllowInvalidTopologySpreadConstraintLabelSelector bool
	// Allow node selector additions for gated pods.
	AllowMutableNodeSelectorAndNodeAffinity bool
	// Allow projected token volumes with non-local paths
	AllowNonLocalProjectedTokenPath bool
	// Allow namespaced sysctls in hostNet and hostIPC pods
	AllowNamespacedSysctlsForHostNetAndHostIPC bool
	// The top-level resource being validated is a Pod, not just a PodSpec
	// embedded in some other resource.
	ResourceIsPod bool
}
```

You can use the feature gate to see an option and just reference that option. This makes it easier for testing and you can allow for skew a bit easier.

Reply:
> Rather than validation code checking the diff and feature gate directly, I would model this as validation options passed into the Validate() and ValidateUpdate() functions. 

OK.  That explains the pattern which I have seen but didn't really look into why we were doing it that way.

Reply:
@thockin 
If we don't verify here, when we use `EnvFrom` in the pod spec, we will allow `configmap/secret` keys that were not allowed before, because `EnvFrom` allows us to directly reference the keys in `configmap/secret` as environment variable names, `configmap/secrets` keys character range  is wider than that of envvar names, the difference is that envvar is not allowed to start with a digit.

Previously, we would verify whether the keys of `configmap/secret` are legal envvar names. If not, we would skip setting this environment variable for the container, output events, and start the container.
ref: https://kubernetes.io/docs/concepts/configuration/secret/#restriction-env-from-invalid

Now we allow all `configmap/secret` keys as envvar names.

According to the [comments](https://github.com/kubernetes/kubernetes/pull/123385#discussion_r1509694339), we shouldn't check the feature gate here (maybe we should), so I modified some documentation for this question, may you take a look.